### PR TITLE
tezos_rpc: add Fetch_block_header

### PR DESCRIPTION
**Pull request by EduardoRFS**
## Depends

- [x] #447

## Problem

Following #447, in the missiong of moving away from Taquito, sometimes we may not receive a block, because the connection with the Tezos RPC was temporarily broken, so we need to be able to fetch old blocks header.

## Solution

Implements `Tezos_rpc.Fetch_block_header`, this will call Tezos asking for an specific block header.

### CODE REVIEW

File: src/tezos_rpc/fetch_block_header.ml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+open Http
+
+type response = Block_header.t
+
+let path ~chain ~block_hash =
+  (* TODO: I don't like this Format.sprintf *)
+  Format.sprintf "/chains/%s/blocks/%s/header" chain block_hash
+
+let execute ~node_uri ~chain ~block_hash =
+  let chain =
```
Comment (gsebil08) [inline src/tezos_rpc/fetch_block_header.ml:11]: In Haskell, this would lead to `shadowing` how is it possible in OCaml? Just because of the `~`?
At first sight, I got confused, is there any OCaml (or Deku) convention to avoid this kind of misleading?
I am used to do something like
```ocaml
  let chain' =
    match chain with
    | Some chain -> Chain_id.to_string chain
    | None -> "main" in
```
Moreover, I am really curious to understand how:
```ocaml
let path = path ~chain ~block_hash in
```
is able to get this `chain` and not the `~chain` provided line 10.

Would you mind help me figure out how this is working? 😄

File: src/tezos_rpc/fetch_block_header.ml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    match chain with
+    | Some chain -> Chain_id.to_string chain
```
Comment (gsebil08) [inline src/tezos_rpc/fetch_block_header.ml:13]: Stupid question, does this "business" belongs to this function? Why don't change `chain_id.t` to a required field, and give the responsibility of giving `main` to the caller?
Please note, once again, I have no clue of the business behind this 😅

File: src/tezos_rpc/fetch_block_header.ml (hunk 1 above)
Comment (EduardoRFS) [reply src/tezos_rpc/fetch_block_header.ml:11]: OCaml values are not recursive by default, unlike Haskell(and most languages), so you can freely shadow variables.

It's the same as the following in Haskell.
```haskell
let x = 1 in
(\x -> print x) (x + 1)
```

Types are recursive by default but you can do non recursive types using `type nonrec t = t`.

File: src/tezos_rpc/fetch_block_header.ml (hunk 2 above)
Comment (EduardoRFS) [reply src/tezos_rpc/fetch_block_header.ml:13]: Chain_id.t cannot be main, this is a special property of this endpoint, it actually accepts `chain_id | "main"`, we could make a variant for that, and I'm still thinking over it, but the easier solution was to do `chain_id option`

**Review by aguillon — APPROVED** [review]

File: src/tezos_rpc/fetch_block_header.ml (hunk 1 above)
Comment (gsebil08) [reply src/tezos_rpc/fetch_block_header.ml:11]: Thank you for the explanation 🙏!

File: src/tezos_rpc/fetch_block_header.ml (hunk 2 above)
Comment (gsebil08) [reply src/tezos_rpc/fetch_block_header.ml:13]: Either I don't understand your answer, or you didn't understand my question 😅.
Here, in `execute` function, we do some basic `match` condition on the `~chain`. If `execute` had been called without `chain` (since it's optional) it will use `main`. Right?

My question is:
Should this function `execute` be responsible of this? In fine, it will be exactly the same. But having `execute ~node_uri chain ~block_hash` can be totally different of having `execute ~node_uri ~chain ~block_hash`.

In my mind, I  want, as fair as possible, my functions to be "functionally scoped". By that, I mean I don't want them to calculate things they are not responsible for.

Comment (gsebil08) [reply src/tezos_rpc/fetch_block_header.ml:13]: Finally got it! Sorry...
